diff --git a/CAESAR/models/caesar_compress.cpp b/CAESAR/models/caesar_compress.cpp
index 55fcb7c..bc7cb35 100644
--- a/CAESAR/models/caesar_compress.cpp
+++ b/CAESAR/models/caesar_compress.cpp
@@ -326,7 +326,7 @@ CompressionResult Compressor::compress(const DatasetConfig& config,
 
       auto t2 = get_start_time();
       std::vector<torch::Tensor> hyper_outputs =
-          hyper_decompressor_model_->run({q_hyper_latent.to(torch::kDouble)});
+          hyper_decompressor_model_->run({q_hyper_latent.to(torch::kFloat32)});
       torch::Tensor mean                 = hyper_outputs[0].to(torch::kFloat32);
       torch::Tensor latent_indexes_recon = hyper_outputs[1].to(torch::kFloat32);
       hyper_outputs.clear();
@@ -358,7 +358,7 @@ CompressionResult Compressor::compress(const DatasetConfig& config,
 
       std::vector<torch::Tensor> decompressor_outputs =
           decompressor_model_->run(
-              {q_latent_with_offset.reshape(new_shape).to(torch::kDouble)});
+              {q_latent_with_offset.reshape(new_shape).to(torch::kFloat32)});
       torch::Tensor raw_output = decompressor_outputs[0].to(torch::kFloat32);
       decompressor_outputs.clear();
 #ifdef USE_CUDA
diff --git a/CAESAR/models/caesar_decompress.cpp b/CAESAR/models/caesar_decompress.cpp
index b4ba224..c338034 100644
--- a/CAESAR/models/caesar_decompress.cpp
+++ b/CAESAR/models/caesar_decompress.cpp
@@ -170,7 +170,7 @@ torch::Tensor Decompressor::decompress(const unsigned int batch_size,
     }
 
     std::vector<torch::Tensor> hyper_outputs = hyper_decompressor_model_->run(
-        {decoded_hyper_latents.to(torch::kDouble).to(device_)});
+        {decoded_hyper_latents.to(torch::kFloat32).to(device_)});
     torch::Tensor mean = hyper_outputs[0].to(torch::kFloat32);
 
     torch::Tensor latent_indexes_recon = hyper_outputs[1].to(torch::kInt32);
@@ -200,7 +200,7 @@ torch::Tensor Decompressor::decompress(const unsigned int batch_size,
     torch::Tensor reshaped_latents = q_latent_with_offset.reshape(new_shape);
 
     std::vector<torch::Tensor> decompressor_outputs =
-        decompressor_model_->run({reshaped_latents.to(torch::kDouble)});
+        decompressor_model_->run({reshaped_latents.to(torch::kFloat32)});
     torch::Tensor raw_output = decompressor_outputs[0].to(torch::kFloat32);
 
 
diff --git a/CAESAR_decompressor.py b/CAESAR_decompressor.py
index dbe621f..b5669e7 100644
--- a/CAESAR_decompressor.py
+++ b/CAESAR_decompressor.py
@@ -470,7 +470,7 @@ state_dict = remove_module_prefix(
     torch.load("./pretrained/caesar_v.pt", map_location=device)
 )
 model.load_state_dict(state_dict)
-model = model.double()
+model = model.float()
 
 quantized_cdf, cdf_length, offset = model.entropy_model.prior._update(30)
 medians = model.entropy_model.prior.medians.detach()
@@ -491,7 +491,7 @@ model.eval()
 with torch.no_grad():
     print("device: ", device)
     model = model.to(device)
-    example_inputs = (torch.randn(10, 2, 64, 16, 16, device=device).double(),)
+    example_inputs = (torch.randn(10, 2, 64, 16, 16, device=device).float(),)
     batch_dim = torch.export.Dim("batch", min=1, max=512)
     # [Optional] Specify the first dimension of the input x as dynamic.
     exported = torch.export.export(
diff --git a/CAESAR_hyper_decompressor.py b/CAESAR_hyper_decompressor.py
index 2178967..e2019b1 100644
--- a/CAESAR_hyper_decompressor.py
+++ b/CAESAR_hyper_decompressor.py
@@ -406,7 +406,7 @@ model = CompressorMix(
 
 state_dict = remove_module_prefix(torch.load('./pretrained/caesar_v.pt', map_location=device))
 model.load_state_dict(state_dict)
-model = model.double()
+model = model.float()
 
 quantized_cdf, cdf_length, offset = model.entropy_model.prior._update(30)
 medians = model.entropy_model.prior.medians.detach()
@@ -421,7 +421,7 @@ model.eval()
 with torch.no_grad():
     print('device: ', device)
     model = model.to(device)
-    example_inputs=(torch.randn(8, 64, 4, 4, device=device).double(),)
+    example_inputs=(torch.randn(8, 64, 4, 4, device=device).float(),)
     batch_dim = torch.export.Dim("batch", min=1, max=1024)
     # [Optional] Specify the first dimension of the input x as dynamic.
     exported = torch.export.export(model, example_inputs, dynamic_shapes={"x": {0: batch_dim}})
