diff --git a/CAESAR_compressor.py b/CAESAR_compressor.py
index 040365c..3a1ae95 100644
--- a/CAESAR_compressor.py
+++ b/CAESAR_compressor.py
@@ -453,7 +453,7 @@ state_dict = remove_module_prefix(
     torch.load("./pretrained/caesar_v.pt", map_location=device, weights_only=False)
 )
 model.load_state_dict(state_dict)
-model = model.half()
+model = model.float()
 
 quantized_cdf, cdf_length, offset = model.entropy_model.prior._update(30)
 medians = model.entropy_model.prior.medians.detach()
@@ -486,7 +486,7 @@ model.eval()
 with torch.no_grad():
     print("device: ", device)
     model = model.to(device)
-    example_inputs = (torch.randn(8, 1, 8, 256, 256, device=device).half(),)
+    example_inputs = (torch.randn(8, 1, 8, 256, 256, device=device).float(),)
     batch_dim = torch.export.Dim("batch", min=1, max=255)
     # [Optional] Specify the first dimension of the input x as dynamic.
     exported = torch.export.export(
